sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -835,5 +896,108 @@ public async ValueTask DisposeAsync()
             await CloseAsync().ConfigureAwait(false);
             GC.SuppressFinalize(this);
         }
+
+        /// <summary>
+        /// Updates the concurrency for the processor. This method can be used to dynamically change the concurrency of a running processor.
+        /// </summary>
+        /// <param name="maxConcurrentCalls">The new max concurrent calls value. This will be reflected in the <see cref="ServiceBusProcessor.MaxConcurrentCalls"/>
+        /// property.</param>
+        public void UpdateConcurrency(int maxConcurrentCalls)
+        {
+            Argument.AssertAtLeast(maxConcurrentCalls, 1, nameof(maxConcurrentCalls));
+            lock (_maxConcurrencySyncLock)
+            {
+                _maxConcurrentCalls = maxConcurrentCalls;
+                // wake up the handler loop
+                _handlerCts.Cancel();
+            }
+        }
+
+        internal void UpdateConcurrency(int maxConcurrentSessions, int maxConcurrentCallsPerSession)
+        {
+            Argument.AssertAtLeast(maxConcurrentSessions, 1, nameof(maxConcurrentSessions));
+            Argument.AssertAtLeast(maxConcurrentCallsPerSession, 1, nameof(maxConcurrentCallsPerSession));
+
+            lock (_maxConcurrencySyncLock)
+            {
+                _maxConcurrentCalls = _sessionIds.Length > 0
+                    ? Math.Min(_sessionIds.Length, maxConcurrentSessions)
+                    : maxConcurrentSessions * maxConcurrentCallsPerSession;
+                _maxConcurrentSessions = maxConcurrentSessions;
+                _maxConcurrentCallsPerSession = maxConcurrentCallsPerSession;
+                // wake up the handler loop
+                _handlerCts.Cancel();
+            }
+        }
+
+        private void ReconcileConcurrency()
+        {
+            EnsureReceiverManagersInitialized();
+
+            int diff = _maxConcurrentCalls - _currentConcurrentCalls;
```

Comment:
read synchronized values once and use them throughout the update. They might change in the middle of the method execution leading to funky results.